pkg/ds/farm.go

```diff
@@ -0,0 +1,411 @@
+					// We should find a nice way to couple them together
+					dsf.logger.NoFields().Infof("Unscheduling from '%s' with dangling daemon set uuid '%s'", nodeName, dsID)
+
+					_, err = dsf.kpStore.DeletePod(kp.INTENT_TREE, nodeName, podID)
+					if err != nil {
+						dsf.logger.NoFields().Errorf("Unable to delete pod id '%v' from intent tree: %v", dsID, err)
+
+						if alertErr := dsf.alerter.Alert(alerting.AlertInfo{
+							Description: fmt.Sprintf("Unable to delete the following pod from store: %v", podID),
+							IncidentKey: "pod_store_delete_error",
+						}); alertErr != nil {
+							dsf.logger.WithError(alertErr).Errorln("Unable to deliver alert!")
+						}
+						continue
+					}
+
+					id := labels.MakePodLabelKey(nodeName, podID)
+					err = dsf.applicator.RemoveLabel(labels.POD, id, DSIDLabel)
+					if err != nil {
+						dsf.logger.NoFields().Errorf("Error removing ds pod id label '%v': %v", id, err)
+					}
+				}
+			}
+		}
+	}()
+}
+
+func (dsf *Farm) mainLoop(quitCh <-chan struct{}) {
+	subQuit := make(chan struct{})
+	defer close(subQuit)
+	dsWatch := dsf.dsStore.Watch(subQuit)
+
+	defer dsf.closeAllChildren()
+
+	var changes dsstore.WatchedDaemonSets
+	var err error
+	var ok bool
+
+	for {
+		select {
+		case <-quitCh:
+			return
+		case changes, ok = <-dsWatch:
+			if !ok {
+				return
+			}
+		}
+
+		// This loop will check all the error channels of the children spawn by this
+		// farm, if any child outputs an error, close the child.
+		// The quitCh here will also return if the caller to mainLoop closes it
+		for dsID, child := range dsf.children {
+			select {
+			case <-quitCh:
+				return
+			case err, ok = <-child.errCh:
+				if !ok {
+					// child error channel closed
+					dsf.closeChild(dsID)
+					continue
+				}
+				dsf.logger.Errorf("An error has occurred in spawned ds '%v':, %v", child.ds, err)
+				if alertErr := dsf.alerter.Alert(alerting.AlertInfo{
+					Description: fmt.Sprintf("An error has occurred in spawned ds '%v':, %v", child.ds, err),
+					IncidentKey: "child_ds_error",
+				}); alertErr != nil {
+					dsf.logger.WithError(alertErr).Errorln("Unable to deliver alert!")
+				}
+				continue
+			default:
+			}
+		}
+		dsf.handleDSChanges(changes, quitCh)
+	}
+}
+
+func (dsf *Farm) closeAllChildren() {
+	for dsID := range dsf.children {
+		dsf.closeChild(dsID)
+	}
+}
+
+func (dsf *Farm) closeChild(dsID fields.ID) {
+	if child, ok := dsf.children[dsID]; ok {
+		close(child.quitCh)
+		close(child.updatedCh)
+		close(child.deletedCh)
+		delete(dsf.children, dsID)
+	}
+}
+
+func (dsf *Farm) handleDSChanges(changes dsstore.WatchedDaemonSets, quitCh <-chan struct{}) {
+	if changes.Err != nil {
+		dsf.logger.Infof("An error has occurred while watching daemon sets: %v", changes.Err)
+		return
+	}
+
+	if len(changes.Created) > 0 {
+		dsf.logger.Infof("The following daemon sets have been created:")
+		for _, dsFields := range changes.Created {
```

Comment:
the daemon set key should be locked somewhere in this function so concurrent farm instances don't try to handle the same daemon set